### PR TITLE
allow for articles to be sorted and ordered by any valid column

### PR DESCRIPTION
**Pull request by kwildeDev**
allow for articles to be sorted by any valid column in ascending or descending order, and handle errors associated with queries at this endpoint

### CODE REVIEW

File: __tests__/app.test.js [hunk 1]
```diff
@@ -163,14 +194,72 @@ describe("/api/articles", () => {
...
+    });
+    it("GET 400: responds with an error message when invalid queries are included alongside valid ones", () => {
+        return request(app)
+        .get("/api/articles?sort_by=topic&hello")
+        .expect(400)
+        .then(({ body }) => {
+            expect(body.msg).toBe("Bad Request");
+        });
+    });
 })
```
Comment (nc-jo-watson) [inline __tests__/app.test.js:263]: Nice Katherine!! I think there is only one test missing - what if I want to use the `sort_by` and `order` queries together?

File: __tests__/app.test.js [hunk 2]
```diff
@@ -163,14 +194,72 @@ describe("/api/articles", () => {
...
+    });
+    it("GET 400: responds with an error message when given an invalid sort_by query", () => {
+        return request(app)
+        .get("/api/articles?sort_by=cats")
+        .expect(400)
+        .then(({ body }) => {
+            expect(body.msg).toBe("Bad Request");
+        });
+    });
+    it("GET 400: responds with an error message when given an invalid order parameter", () => {
```
Comment (nc-jo-watson) [inline __tests__/app.test.js:247]: Nitpick! this is a query not a parameter

File: controllers/articles.controller.js [hunk 3]
```diff
@@ -13,7 +13,11 @@ exports.getArticleById = (request, response, next) => {
 }
 
 exports.getArticles = (request, response, next) => {
-    fetchArticles()
+    const queryKeys = Object.keys(request.query)
+    const invalidKeys = queryKeys.filter((key) => key !== 'sort_by' && key !== 'order')
+    const sort_by = invalidKeys.length ? "invalid_query" : request.query.sort_by
```
Comment (nc-jo-watson) [inline controllers/articles.controller.js:18]: It's not clear to me what this is for. If your endpoint is passed queries that aren't handled, we can just ignore them. When you extract the queries you want from req.queries (like you've done on the line below this comment!), we implicitly disregard any other queries.

File: models/articles.models.js [hunk 4]
```diff
@@ -11,7 +11,12 @@ exports.fetchArticleById = (article_id) => {
     });
 }
 
-exports.fetchArticles = () => {
+exports.fetchArticles = (sort_by = "created_at", order = "desc") => {
+    const validSortBys = ["author", "title", "topic", "created_at", "votes", "comment_count"];
+    const validOrders = ["asc", "desc"]
+    if (!validSortBys.includes(sort_by) || (!validOrders.includes(order))) {
+        return Promise.reject({ status: 400, msg: "Bad Request"})
```
Comment (nc-jo-watson) [inline models/articles.models.js:18]: Good to see green listing! 💚

**Review by nc-jo-watson — COMMENTED** [review]
Really nice Katherine! 😍  The only question I had was about how you handle the queries in your controller.

File: controllers/articles.controller.js (hunk 3 above)
Comment (kwildeDev) [reply controllers/articles.controller.js:18]: Thank you, I see your point. I have sorted this out now!

**Pull request merged**